File: 133-clone-graph/133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""
from collections import deque
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        queue = []
        queue.append(node)
        old2new = dict()
        if not node:
            return node
        newNode = Node(node.val, [])
        old2new[node] = newNode
        while(queue):
            size = len(queue)
            for i in range(size):
                popped = queue.pop(0)
                for nxt in popped.neighbors:
                    if nxt in old2new:
                        old2new[popped].neighbors.append(old2new[nxt])
                        continue
                    newNode = Node(nxt.val, [])
                    old2new[nxt] = newNode
                    old2new[popped].neighbors.append(old2new[nxt])
                    queue.append(nxt)
        return old2new[node]
                    
        
        

File: 133-clone-graph/133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        //@input : one node denoting a connected graph
        // in a node, there is a value and a adjacency list for neighbors
        //@output : a cloned graph
        //@edge case : no
        //@breaking down:
        // 1. bfs, if not copy, create the copy of this node and add into the copied values;
        // 2. traverse the adjacency list and if uncopied, then create a copy and add into the adjacency list; if copied, then add copied node into the adjacency list.
       //3. add the node into the queue
        // 4. repeat the process until queue is empty
        LinkedList<Node> queue = new LinkedList<>();
        Map<Node, Node> copyToOriginal = new HashMap<>();
        Map<Node, Node> originalToCopy = new HashMap<>();
        if(node == null)return null;
        Node root = new Node(node.val);
        originalToCopy.put(node, root);
        copyToOriginal.put(root, node);
        queue.addLast(root);
        Set<Node> visited = new HashSet<>();
        visited.add(root);
        //bfs，注意不要重复
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
                Node popped = queue.pollFirst();
                Node poppedOriginal = copyToOriginal.get(popped);
                for(Node neighbor: poppedOriginal.neighbors){
                    if(!originalToCopy.containsKey(neighbor)){
                        Node copiedNeighbor = new Node(neighbor.val);
                        originalToCopy.put(neighbor, copiedNeighbor);
                        copyToOriginal.put(copiedNeighbor, neighbor);
                    }
                    if(!visited.contains(originalToCopy.get(neighbor))){
                        queue.addLast(originalToCopy.get(neighbor));
                        visited.add(originalToCopy.get(neighbor));
                    }
                    popped.neighbors.add(originalToCopy.get(neighbor));
                }
                
            }
            
        }
        //System.out.println(node.val);
        return root;
        
        
        
        
        
    }
}